### PR TITLE
ADR on logging

### PR DESCRIPTION
**Pull request by withinfocus**
## Objective

ADR on refactoring of logging. Relates to #155. Successor of #151.

### CODE REVIEW

File: docs/architecture/adr/0021-console-logging.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
+---
+adr: "0021"
+status: In progress
+date: 2023-07-13
+tags: [server]
+---
+
+# 0021 - Logging with the Console
```
Comment (Hinton) [inline docs/architecture/adr/0021-console-logging.md:8]: While console is well termed in the dotnet world due to Windows calling it "Console", it's typically terminal in the linux world. Stdout is a bit more generic.

```suggestion
# 0021 - Logging to stdout
```

File: docs/architecture/adr/0021-console-logging.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+added as dependencies leaving this core logging layer with a growing list of conditions on how and
+when to use certain types of structured logging. Maintenance needs have grown to keep sink
+dependencies up to date and more are desired to be added. Some of the presently-available sinks have
+very little use and / or better alternatives now exist.
+
+## Considered Options
+
+:::note
+
+Bitwarden uses [Datadog][dd] as its monitoring tool and desires to increase its usage by engineers
```
Comment (Hinton) [inline docs/architecture/adr/0021-console-logging.md:26]:
```suggestion
Bitwarden currently uses [Datadog][dd] as its monitoring tool and desires to increase its usage by engineers
```

File: docs/architecture/adr/0021-console-logging.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+## Context and Problem Statement
+
+As the server platform has matured so have the various logging extensions to support additional use
+cases and customer requests. [Serilog][serilog] is in place via shared logic for all services,
+initiated at startup, and over time additional its "sinks" with their needed configuration have been
+added as dependencies leaving this core logging layer with a growing list of conditions on how and
+when to use certain types of structured logging. Maintenance needs have grown to keep sink
+dependencies up to date and more are desired to be added. Some of the presently-available sinks have
+very little use and / or better alternatives now exist.
```
Comment (Hinton) [inline docs/architecture/adr/0021-console-logging.md:20]: I think this section is a bit hard to follow.

Thoughts on splitting it into two paragraphs, starting with how the current logging solution works, and ending with why we want to change it. Something like.

> Bitwarden uses Serilog with both standard and custom logic to handle multiple outputs ....
> 
> Maintaining sinks that are never used is annoying .... and custom logic for file rotation ...

File: docs/architecture/adr/0021-console-logging.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+across the board to improve what we deliver.
+
+:::
+
+- **Maintain current logging options** - Support what is available today for logging methods and
+  expect those running the platform to configure what they need outside of it for log collection.
+- **Extend the plaform to specifically support Datadog** - A Serilog sink [exists][ddsink] and the
+  platform can send logs directly to Datadog.
+- **Consolidate logging providers** - Announce deprecation and migration plans for sinks not aligned
+  with core needs and center on the console for logs.
```
Comment (Hinton) [inline docs/architecture/adr/0021-console-logging.md:36]:
```suggestion
- **Consolidate logging providers** - Announce deprecation and migration plans for sinks not aligned
  with core needs and center on the stdout for logs.
```

Comment (Hinton) [reply docs/architecture/adr/0021-console-logging.md:36]: I might even got a step further and call this option "Adopt cloud native logging practices - Deprecate existing sinks and align on stdout & file based logging."

File: docs/architecture/adr/0021-console-logging.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+### Plan
+
+Using standard support policies, release notes will include a mention that three Serilog sinks will
+be removed:
+
+- CosmosDb
+- Sentry
+- Syslog
+
+The remaining sinks -- core functionality of Serilog -- will continue to be supported:
```
Comment (Hinton) [inline docs/architecture/adr/0021-console-logging.md:65]: Are we motivating the use of Serilog over .net core logging anywhere?

File: docs/architecture/adr/0021-console-logging.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+- Console
+- File
+
+While the Serilog [console sink][serilogconsole] is currently an implicit dependency with what's
+provided for ASP.NET Core, it will be explicitly referenced.
+
+Solutions exist for users to shift processing of logs for the removed sinks to console or file and
+retain their integration. Admin Portal users can similarly continue to use CosmosDb for log
+retention, but it is suggested that application monitoring that's available be used instead that
+should in essentially all cases be able to receive and process console logs.
```
Comment (Hinton) [inline docs/architecture/adr/0021-console-logging.md:76]: I inferred that this work would also remove the CosmosDb sink is that not correct?

**Review by Hinton — COMMENTED** [review]
Looks good overall, some thoughts.

File: docs/architecture/adr/0021-console-logging.md (hunk 5 above)
Comment (withinfocus) [reply docs/architecture/adr/0021-console-logging.md:65]: Not sure what you mean here -- as in make a case for Serilog instead of MEL?

File: docs/architecture/adr/0021-console-logging.md (hunk 6 above)
Comment (withinfocus) [reply docs/architecture/adr/0021-console-logging.md:76]: Yes -- line 61 here.

File: docs/architecture/adr/0021-console-logging.md (hunk 5 above)
Comment (Hinton) [reply docs/architecture/adr/0021-console-logging.md]: Yes. The ADR mentions being able to use MEL. We should write down why we decided to use Serilog instead.

File: docs/architecture/adr/0021-standard-output-logging.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+This will allow better usage of `appsettings.json` and a richer developer experience. Existing
+built-in [.NET Core logging][netcorelogging] will continue to be available if desired, but the
+recommendation will be to move to a Serilog configuration.
+
+Code cleanup will be performed around Serilog usage, such as:
+
+- Removal of overuse of inclusion predicates that complicate (or sometimes block) effective log
+  output.
+- Alignment with .NET Core and Serilog best practices on initialization and usage of Serilog itself.
```
Comment (Hinton) [inline docs/architecture/adr/0021-standard-output-logging.md:108]: Could we link an example here.

File: docs/architecture/adr/0021-standard-output-logging.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+- Improvements in logging initialization reliability and working with configuration issues, as well
+  as more resilient tear-down when a component stops / ends.
```
Comment (Hinton) [inline docs/architecture/adr/0021-standard-output-logging.md:110]: Do you have any more context on what this involves?

File: docs/architecture/adr/0021-standard-output-logging.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+- Removal of the above-deprecated sinks, in the final release of the support window.
+
+Logging functionality will be moved to a new shared library -- separate from the core -- for
```
Comment (Hinton) [inline docs/architecture/adr/0021-standard-output-logging.md:113]:
```suggestion
Logging functionality will be moved to a new shared library -- separate from the core package -- for
```

**Review by Hinton — COMMENTED** [review]
Some minor comments.
